File: euler/python/P4.py
"""
Largest Palindrome Product
Problem 4
"""

MIN, MAX = 100, 999

if __name__ == '__main__':
    # Generate a list of all products of two three-digit numbers
    numbers = [i * j for i in range(MIN, MAX) for j in range(MIN, i)]
    # Filter out non-palindromes and print the maximum
    numbers = filter(lambda x: str(x) == str(x)[::-1], numbers)
    print max(numbers)

File: euler/ruby/P4.rb
"
Problem 4
Largest Palindrome Product
"

NUMBER_BASE = 10
NUM_OF_DIGITS = 3
PRODUCT_DEGREE = 2

# Find the range of numbers we want to work in (the maximum is exclusive)
minimum = NUMBER_BASE ** (NUM_OF_DIGITS - 1)
maximum = minimum * NUMBER_BASE

puts (minimum...maximum)
  .to_a.combination(PRODUCT_DEGREE)        # Find combinations of two numbers
  .map { |n| n.inject(:*) }                # Multiply the numbers together
  .select { |n| n.to_s.reverse.to_i == n } # Filter non-palindrome multiples
  .max                                     # Find the maximum
